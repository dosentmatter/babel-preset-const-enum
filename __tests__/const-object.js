import { transformAsync } from '@babel/core';
import preset from '../src';

const options = {
  filename: 'file.ts',
  presets: [[preset, { transform: 'constObject' }]],
};

it('Transforms no initializers', async () => {
  const input = `const enum Direction { Left, Right, Down, Up };
`;

  const { code: output } = await transformAsync(input, options);
  expect(output).toMatchSnapshot();
});

it('Transforms string members', async () => {
  const input = `const enum Enum {
  A = 1,
  B = A,
  C = '',
  D = C,
  E = 1,
  F,
}
`;

  const { code: output } = await transformAsync(input, options);
  expect(output).toMatchSnapshot();
});

it('Transforms computed members', async () => {
  const input = `const enum MyEnum {
  A = 1,
  B = A,
  C,
  D = C,
  E = 1,
  F,
  G = A * E,
  H = A ** B ** C,
  I = A << 20
}
`;

  const { code: output } = await transformAsync(input, options);
  expect(output).toMatchSnapshot();

  const MyEnum = new Function(
    `${output}
return MyEnum;
`,
  )();
  expect(MyEnum.A).toBe(1);
  expect(MyEnum.B).toBe(1);
  expect(MyEnum.C).toBe(2);
  expect(MyEnum.D).toBe(2);
  expect(MyEnum.E).toBe(1);
  expect(MyEnum.F).toBe(2);
  expect(MyEnum.G).toBe(1);
  expect(MyEnum.H).toBe(1);
  expect(MyEnum.I).toBe(1048576);
});

it('Transforms chained computed members', async () => {
  const input = `const enum MyEnum {
  A = 1,
  B = A * 2,
  C,
  D = C,
  E = D ** 2,
  F,
  G = F * E,
  H,
  I = H << 20
}
`;

  const { code: output } = await transformAsync(input, options);
  expect(output).toMatchSnapshot();

  const MyEnum = new Function(
    `${output}
return MyEnum;
`,
  )();
  expect(MyEnum.A).toBe(1);
  expect(MyEnum.B).toBe(2);
  expect(MyEnum.C).toBe(3);
  expect(MyEnum.D).toBe(3);
  expect(MyEnum.E).toBe(9);
  expect(MyEnum.F).toBe(10);
  expect(MyEnum.G).toBe(90);
  expect(MyEnum.H).toBe(91);
  expect(MyEnum.I).toBe(95420416);
});
