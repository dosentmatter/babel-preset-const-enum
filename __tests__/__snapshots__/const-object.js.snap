// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transforms chained computed members 1`] = `
"\\"use strict\\";

const MyEnum = {
  A: 1,
  B: 2,
  C: 3,
  D: 3,
  E: 9,
  F: 10,
  G: 90,
  H: 91,
  I: 95420416
};"
`;

exports[`Transforms computed members 1`] = `
"\\"use strict\\";

const MyEnum = {
  A: 1,
  B: 1,
  C: 2,
  D: 2,
  E: 1,
  F: 2,
  G: 1,
  H: 1,
  I: 1048576
};"
`;

exports[`Transforms no initializers 1`] = `
"\\"use strict\\";

const Direction = {
  Left: 0,
  Right: 1,
  Down: 2,
  Up: 3
};
;"
`;

exports[`Transforms string members 1`] = `
"\\"use strict\\";

const Enum = {
  A: 1,
  B: 1,
  C: \\"\\",
  D: \\"\\",
  E: 1,
  F: 2
};"
`;
